File: lib/cadence_web/live/ops_console_live/widgets/procedure_widget.ex
defmodule CadenceWeb.OpsConsoleLive.Widgets.ProcedureWidget do
  @moduledoc """
  Phoenix LiveComponent for displaying procedure execution status.

  Shows:
  - Progress bar with step count
  - Step list with status icons
  - Elapsed time (via JavaScript hook)
  - Control buttons (pause/resume/abort)
  - Log tail

  ## Usage

      <.live_component
        module={ProcedureWidget}
        id="proc-widget-1"
        widget_id="widget-def"
        execution={@execution}
        config={%{
          "show_logs" => true,
          "max_log_lines" => 10
        }}
        on_configure={JS.push("open_widget_config")}
        on_pause={JS.push("pause_procedure")}
        on_resume={JS.push("resume_procedure")}
        on_abort={JS.push("abort_procedure")}
      />
  """
  use CadenceWeb, :live_component

  import CadenceWeb.WidgetComponents

  alias Phoenix.LiveView.JS

  @impl true
  def update(assigns, socket) do
    {:ok, assign(socket, assigns)}
  end

  @impl true
  def render(assigns) do
    config = assigns[:config] || %{}
    execution = assigns[:execution]
    title = config["title"] || (execution && execution.procedure_name) || "Procedure"
    show_logs = config["show_logs"] != false
    max_log_lines = config["max_log_lines"] || 10

    # Calculate progress
    {current_step, total_steps, progress_pct} = calculate_progress(execution)

    assigns =
      assigns
      |> assign(:title, title)
      |> assign(:show_logs, show_logs)
      |> assign(:max_log_lines, max_log_lines)
      |> assign(:current_step, current_step)
      |> assign(:total_steps, total_steps)
      |> assign(:progress_pct, progress_pct)

    ~H"""
    <.widget
      id={@widget_id}
      title={@title}
      on_configure={
        @on_configure &&
          JS.push("open_widget_config", value: widget_config_value(@widget_id, "procedure", @config))
      }
    >
      <div class="procedure-widget">
        <%= if @execution do %>
          <!-- Progress header -->
          <div class="procedure-header mb-3">
            <div class="flex items-center justify-between mb-2">
              <div class="flex items-center gap-2">
                <.severity_dot
                  severity={status_severity(@execution.status)}
                  pulse={@execution.status == :running}
                />
                <span class="text-sm font-medium">{format_status(@execution.status)}</span>
              </div>
              <div
                id={"#{@widget_id}-elapsed"}
                class="text-xs text-base-content/50 font-mono"
                phx-hook="ProcedureElapsed"
                data-started-at={@execution.started_at && DateTime.to_iso8601(@execution.started_at)}
                data-status={@execution.status}
              >
                --:--
              </div>
            </div>
            
    <!-- Progress bar -->
            <div class="w-full bg-base-300 rounded-full h-1.5 mb-1">
              <div
                class="h-1.5 rounded-full bg-primary transition-all"
                style={"width: #{@progress_pct}%"}
              >
              </div>
            </div>
            <div class="text-xs text-base-content/50">
              Step {@current_step} of {@total_steps}
            </div>
          </div>
          
    <!-- Control buttons -->
          <div class="procedure-controls flex items-center gap-2 mb-3">
            <.widget_action_btn
              :if={@execution.status == :running && @on_pause}
              phx-click={JS.push("pause_procedure", value: %{id: @execution.id})}
            >
              <.icon name="hero-pause" class="w-3 h-3" /> Pause
            </.widget_action_btn>

            <.widget_action_btn
              :if={@execution.status == :paused && @on_resume}
              variant="success"
              phx-click={JS.push("resume_procedure", value: %{id: @execution.id})}
            >
              <.icon name="hero-play" class="w-3 h-3" /> Resume
            </.widget_action_btn>

            <.widget_action_btn
              :if={@execution.status in [:running, :paused] && @on_abort}
              variant="danger"
              phx-click={JS.push("abort_procedure", value: %{id: @execution.id})}
            >
              <.icon name="hero-stop" class="w-3 h-3" /> Abort
            </.widget_action_btn>
          </div>
          
    <!-- Step list -->
          <div class="procedure-steps flex-1 overflow-y-auto space-y-1 mb-3">
            <div
              :for={{step, idx} <- Enum.with_index(@execution.steps || [])}
              class={[
                "step-item flex items-center gap-2 p-1.5 rounded text-sm",
                step_bg_class(step, idx, @current_step - 1)
              ]}
            >
              <div class="step-indicator">
                {step_icon(step, idx, @current_step - 1)}
              </div>
              <span class="truncate flex-1">{step.name || "Step #{idx + 1}"}</span>
            </div>
          </div>
          
    <!-- Log tail -->
          <div
            :if={@show_logs && @execution.logs && length(@execution.logs) > 0}
            class="procedure-logs border-t border-base-300 pt-2"
          >
            <div class="text-xs text-base-content/40 mb-1">Recent logs</div>
            <div class="space-y-0.5 font-mono text-xs">
              <div
                :for={log <- Enum.take(@execution.logs, -@max_log_lines)}
                class={["log-line", log_level_class(log.level)]}
              >
                <span class="text-base-content/40">{format_log_time(log.timestamp)}</span>
                <span class="ml-2">{log.message}</span>
              </div>
            </div>
          </div>
        <% else %>
          <.widget_empty icon="hero-document-text" message="No procedure running" />
        <% end %>
      </div>
    </.widget>
    """
  end

  defp calculate_progress(nil), do: {0, 0, 0}

  defp calculate_progress(execution) do
    steps = execution.steps || []
    total = length(steps)

    if total == 0 do
      {0, 0, 0}
    else
      current = (execution.current_step_index || 0) + 1
      pct = round(current / total * 100)
      {current, total, pct}
    end
  end

  defp status_severity(:running), do: :info
  defp status_severity(:paused), do: :warning
  defp status_severity(:completed), do: :nominal
  defp status_severity(:failed), do: :critical
  defp status_severity(:cancelled), do: :critical
  defp status_severity(_), do: :info

  defp format_status(:running), do: "Running"
  defp format_status(:paused), do: "Paused"
  defp format_status(:completed), do: "Completed"
  defp format_status(:failed), do: "Failed"
  defp format_status(:cancelled), do: "Cancelled"
  defp format_status(:pending), do: "Pending"
  defp format_status(status), do: to_string(status)

  defp step_bg_class(_step, idx, current_idx) when idx < current_idx, do: "bg-success/10"
  defp step_bg_class(_step, idx, current_idx) when idx == current_idx, do: "bg-primary/10"
  defp step_bg_class(_step, _idx, _current_idx), do: ""

  defp step_icon(_step, idx, current_idx) when idx < current_idx do
    assigns = %{}

    ~H"""
    <.icon name="hero-check-circle" class="w-4 h-4 text-success" />
    """
  end

  defp step_icon(_step, idx, current_idx) when idx == current_idx do
    assigns = %{}

    ~H"""
    <.icon name="hero-arrow-right-circle" class="w-4 h-4 text-primary animate-pulse" />
    """
  end

  defp step_icon(_step, _idx, _current_idx) do
    assigns = %{}

    ~H"""
    <.icon name="hero-circle-stack" class="w-4 h-4 text-base-content/30" />
    """
  end

  defp log_level_class(:error), do: "text-error"
  defp log_level_class(:warning), do: "text-warning"
  defp log_level_class(:info), do: "text-info"
  defp log_level_class(_), do: "text-base-content/70"

  defp format_log_time(nil), do: "--:--:--"

  defp format_log_time(%DateTime{} = dt) do
    Calendar.strftime(dt, "%H:%M:%S")
  end

  defp format_log_time(_), do: "--:--:--"

  defp widget_config_value(widget_id, widget_type, config) do
    %{widget_id: widget_id, widget_type: widget_type, config: config}
  end
end

File: legacy/cadence_legacy/assets/js/widgets/procedure_widget.js
/**
 * Procedure Status Widget
 *
 * Shows running procedure with real-time progress.
 * Features:
 * - Step list with completion states (pending/running/completed/failed/skipped)
 * - Current step highlight with elapsed time
 * - Pause/Resume/Abort controls
 * - Progress bar
 * - Real-time log tail (last few lines)
 */

import { BaseWidget, registerWidget } from "../dashboard/widget_factory"

/**
 * ProcedureWidget - Procedure execution status and control.
 */
export class ProcedureWidget extends BaseWidget {
  static displayName = "Procedure Status"
  static description = "Monitor and control running procedure execution"
  static icon = "clipboard-check"

  constructor(options) {
    super(options)
    this.type = "procedure"
    this.execution = null
    this.steps = []
    this.logs = []
    this.elapsedTimer = null
  }

  init() {
    this.render()
    this._startElapsedTimer()
    this.initialized = true
  }

  renderContent() {
    const {
      procedure_name,
      execution_id,
      show_logs = true,
      max_visible_steps = 5
    } = this.config

    // Mock execution data for display (in real implementation, this comes from LiveView)
    const execution = this.execution || {
      status: 'idle',
      current_step: null,
      started_at: null,
      completed_steps: 0,
      total_steps: 0
    }

    const steps = this.steps || []
    const logs = this.logs || []

    if (!procedure_name && !execution_id) {
      return `
        <div class="procedure-widget flex items-center justify-center h-full">
          <div class="text-center text-base-content/50">
            <svg class="w-8 h-8 mx-auto mb-2 opacity-50" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 5H7a2 2 0 00-2 2v10a2 2 0 002 2h8a2 2 0 002-2V7a2 2 0 00-2-2h-2M9 5a2 2 0 002 2h2a2 2 0 002-2M9 5a2 2 0 012-2h2a2 2 0 012 2"/>
            </svg>
            <p class="text-sm">No procedure configured</p>
          </div>
        </div>
      `
    }

    const progressPercent = execution.total_steps > 0
      ? Math.round((execution.completed_steps / execution.total_steps) * 100)
      : 0

    return `
      <div class="procedure-widget flex flex-col h-full">
        <!-- Header with procedure name and controls -->
        <div class="procedure-header flex items-center justify-between mb-2">
          <div class="flex-1 min-w-0">
            <div class="font-semibold text-sm truncate">${procedure_name || 'Procedure'}</div>
            <div class="text-xs text-base-content/50">${this._getStatusLabel(execution.status)}</div>
          </div>
          <div class="procedure-controls flex items-center gap-1">
            ${this._renderControls(execution.status)}
          </div>
        </div>

        <!-- Progress bar -->
        <div class="procedure-progress mb-3">
          <div class="flex items-center justify-between text-xs text-base-content/50 mb-1">
            <span>Step ${execution.completed_steps}/${execution.total_steps}</span>
            <span>${progressPercent}%</span>
          </div>
          <div class="w-full h-2 bg-base-300 rounded-full overflow-hidden">
            <div
              class="h-full ${this._getProgressBarClass(execution.status)} transition-all duration-300"
              style="width: ${progressPercent}%"
            ></div>
          </div>
        </div>

        <!-- Steps list -->
        <div class="procedure-steps flex-1 overflow-y-auto mb-2">
          ${this._renderSteps(steps, execution.current_step, max_visible_steps)}
        </div>

        <!-- Log tail -->
        ${show_logs ? `
          <div class="procedure-logs mt-auto pt-2 border-t border-base-300">
            <div class="mc-label-subsystem text-base-content/40 mb-1">LOG</div>
            <div class="log-content bg-base-300/50 rounded p-2 h-16 overflow-y-auto font-mono text-xs">
              ${logs.length > 0 ? logs.slice(-5).map(log => `
                <div class="log-line ${this._getLogClass(log.level)}">
                  <span class="text-base-content/40">${this._formatLogTime(log.timestamp)}</span>
                  <span>${log.message}</span>
                </div>
              `).join('') : '<span class="text-base-content/40">No log entries</span>'}
            </div>
          </div>
        ` : ''}

        <!-- Elapsed time -->
        ${execution.started_at ? `
          <div class="procedure-elapsed text-center mt-2">
            <span class="text-xs text-base-content/40">Elapsed: </span>
            <span class="text-xs font-mono" id="elapsed-${this.id}">${this._formatElapsed(execution.started_at)}</span>
          </div>
        ` : ''}
      </div>
    `
  }

  _renderControls(status) {
    switch (status) {
      case 'running':
        return `
          <button class="btn btn-ghost btn-xs" id="proc-pause-${this.id}" title="Pause">
            <svg class="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M10 9v6m4-6v6m7-3a9 9 0 11-18 0 9 9 0 0118 0z"/>
            </svg>
          </button>
          <button class="btn btn-ghost btn-xs text-error" id="proc-abort-${this.id}" title="Abort">
            <svg class="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"/>
            </svg>
          </button>
        `
      case 'paused':
        return `
          <button class="btn btn-ghost btn-xs text-success" id="proc-resume-${this.id}" title="Resume">
            <svg class="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M14.752 11.168l-3.197-2.132A1 1 0 0010 9.87v4.263a1 1 0 001.555.832l3.197-2.132a1 1 0 000-1.664z"/>
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z"/>
            </svg>
          </button>
          <button class="btn btn-ghost btn-xs text-error" id="proc-abort-${this.id}" title="Abort">
            <svg class="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"/>
            </svg>
          </button>
        `
      case 'idle':
        return `
          <button class="btn btn-ghost btn-xs text-success" id="proc-start-${this.id}" title="Start">
            <svg class="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M14.752 11.168l-3.197-2.132A1 1 0 0010 9.87v4.263a1 1 0 001.555.832l3.197-2.132a1 1 0 000-1.664z"/>
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z"/>
            </svg>
          </button>
        `
      default:
        return ''
    }
  }

  _renderSteps(steps, currentStepId, maxVisible) {
    if (steps.length === 0) {
      return '<div class="text-xs text-base-content/40 text-center py-2">No steps loaded</div>'
    }

    // Find current step index to show context around it
    const currentIndex = steps.findIndex(s => s.id === currentStepId)
    let visibleSteps = steps

    if (steps.length > maxVisible && currentIndex >= 0) {
      // Show steps around the current one
      const start = Math.max(0, currentIndex - Math.floor(maxVisible / 2))
      const end = Math.min(steps.length, start + maxVisible)
      visibleSteps = steps.slice(start, end)
    } else if (steps.length > maxVisible) {
      visibleSteps = steps.slice(0, maxVisible)
    }

    return visibleSteps.map((step, idx) => `
      <div class="step-item flex items-start gap-2 py-1 ${step.id === currentStepId ? 'bg-primary/10 -mx-2 px-2 rounded' : ''}">
        <div class="step-indicator flex-shrink-0 mt-0.5">
          ${this._getStepIcon(step.status, step.id === currentStepId)}
        </div>
        <div class="step-content flex-1 min-w-0">
          <div class="flex items-center gap-2">
            <span class="text-xs font-mono text-base-content/40">${step.number || idx + 1}.</span>
            <span class="text-sm truncate ${step.status === 'failed' ? 'text-error' : ''}">${step.name}</span>
          </div>
          ${step.duration ? `
            <span class="text-xs text-base-content/40">${this._formatDuration(step.duration)}</span>
          ` : ''}
        </div>
      </div>
    `).join('')
  }

  _getStepIcon(status, isCurrent) {
    const iconClass = "w-4 h-4"

    switch (status) {
      case 'completed':
        return `<svg class="${iconClass} text-success" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 13l4 4L19 7"/>
        </svg>`
      case 'running':
        return `<span class="${iconClass} flex items-center justify-center">
          <span class="loading loading-spinner loading-xs text-primary"></span>
        </span>`
      case 'failed':
        return `<svg class="${iconClass} text-error" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"/>
        </svg>`
      case 'skipped':
        return `<svg class="${iconClass} text-base-content/30" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M13 5l7 7-7 7M5 5l7 7-7 7"/>
        </svg>`
      default: // pending
        return `<span class="w-3 h-3 rounded-full border-2 border-base-content/20 ${isCurrent ? 'border-primary' : ''}"></span>`
    }
  }

  _getStatusLabel(status) {
    const labels = {
      idle: 'Ready to run',
      running: 'Running',
      paused: 'Paused',
      completed: 'Completed',
      failed: 'Failed',
      aborted: 'Aborted'
    }
    return labels[status] || 'Unknown'
  }

  _getProgressBarClass(status) {
    switch (status) {
      case 'running':
        return 'bg-primary'
      case 'paused':
        return 'bg-warning'
      case 'completed':
        return 'bg-success'
      case 'failed':
      case 'aborted':
        return 'bg-error'
      default:
        return 'bg-base-content/20'
    }
  }

  _getLogClass(level) {
    switch (level) {
      case 'error':
        return 'text-error'
      case 'warning':
        return 'text-warning'
      case 'success':
        return 'text-success'
      default:
        return 'text-base-content/70'
    }
  }

  _formatLogTime(timestamp) {
    if (!timestamp) return ''
    const date = new Date(timestamp)
    return date.toLocaleTimeString('en-US', {
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: false
    })
  }

  _formatElapsed(startedAt) {
    if (!startedAt) return '00:00:00'

    const start = new Date(startedAt)
    const now = new Date()
    const diff = Math.floor((now - start) / 1000)

    const hours = Math.floor(diff / 3600)
    const minutes = Math.floor((diff % 3600) / 60)
    const seconds = diff % 60

    return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`
  }

  _formatDuration(ms) {
    if (!ms) return ''
    const seconds = Math.round(ms / 1000)
    if (seconds < 60) return `${seconds}s`
    const minutes = Math.floor(seconds / 60)
    const remainingSeconds = seconds % 60
    return `${minutes}m ${remainingSeconds}s`
  }

  _startElapsedTimer() {
    // Update elapsed time every second
    this.elapsedTimer = setInterval(() => {
      const elapsedEl = this.container.querySelector(`#elapsed-${this.id}`)
      if (elapsedEl && this.execution?.started_at) {
        elapsedEl.textContent = this._formatElapsed(this.execution.started_at)
      }
    }, 1000)
  }

  render() {
    super.render()
    this._bindControls()
  }

  _bindControls() {
    const pauseBtn = this.container.querySelector(`#proc-pause-${this.id}`)
    const resumeBtn = this.container.querySelector(`#proc-resume-${this.id}`)
    const abortBtn = this.container.querySelector(`#proc-abort-${this.id}`)
    const startBtn = this.container.querySelector(`#proc-start-${this.id}`)

    pauseBtn?.addEventListener('click', (e) => {
      e.stopPropagation()
      this._dispatchAction('pause')
    })

    resumeBtn?.addEventListener('click', (e) => {
      e.stopPropagation()
      this._dispatchAction('resume')
    })

    abortBtn?.addEventListener('click', (e) => {
      e.stopPropagation()
      this._dispatchAction('abort')
    })

    startBtn?.addEventListener('click', (e) => {
      e.stopPropagation()
      this._dispatchAction('start')
    })
  }

  _dispatchAction(action) {
    const event = new CustomEvent("widget:procedure_action", {
      detail: {
        widgetId: this.id,
        executionId: this.config.execution_id,
        action: action
      },
      bubbles: true
    })
    this.container.dispatchEvent(event)
  }

  // Called by external code to update execution state
  updateExecution(execution, steps, logs) {
    this.execution = execution
    this.steps = steps || this.steps
    this.logs = logs || this.logs
    this.render()
  }

  // Add a log entry
  addLog(log) {
    this.logs.push(log)
    // Keep only last 50 logs
    if (this.logs.length > 50) {
      this.logs = this.logs.slice(-50)
    }
    this.render()
  }

  openConfig() {
    const event = new CustomEvent("widget:configure", {
      detail: {
        widgetId: this.id,
        widgetType: this.type,
        config: this.getConfig()
      },
      bubbles: true
    })
    this.container.dispatchEvent(event)
  }

  destroy() {
    if (this.elapsedTimer) {
      clearInterval(this.elapsedTimer)
    }
    super.destroy()
  }
}

// Register the widget
registerWidget("procedure", ProcedureWidget)

export default ProcedureWidget
